allennlp/common/from_params.py

```diff
@@ -0,0 +1,185 @@
+        elif hasattr(annotation, 'from_params'):
+            if name in params:
+                # Our params have an entry for this, so we use that.
+                subparams = params.pop(name)
+
+                if takes_arg(annotation.from_params, 'extras'):
+                    # If annotation.params accepts **extras, we need to pass them all along.
+                    subextras = extras
+                else:
+                    # Otherwise, only supply the ones that are actual args; any additional ones
+                    # will cause a TypeError.
+                    subextras = {k: v for k, v in extras.items() if takes_arg(annotation.from_params, k)}
+
+                # In some cases we allow a string instead of a param dict, so
+                # we need to handle that case separately.
+                if isinstance(subparams, str):
+                    kwargs[name] = annotation.by_name(subparams)()
+                else:
+                    kwargs[name] = annotation.from_params(params=subparams, **subextras)
+            elif not optional:
+                # Not optional and not supplied, that's an error!
+                raise ConfigurationError(f"expected key {name} for {cls.__name__}")
+
+        # If the parameter type is a Python primitive, just pop it off
+        # using the correct casting pop_xyz operation.
+        elif annotation == str:
+            kwargs[name] = (params.pop(name, default)
+                            if optional
+                            else params.pop(name))
+        elif annotation == int:
+            kwargs[name] = (params.pop_int(name, default)
+                            if optional
+                            else params.pop_int(name))
+        elif annotation == bool:
+            kwargs[name] = (params.pop_bool(name, default)
+                            if optional
+                            else params.pop_bool(name))
+        elif annotation == float:
+            kwargs[name] = (params.pop_float(name, default)
+                            if optional
+                            else params.pop_float(name))
+
+        # This handles types like Dict[str, TokenIndexer], where we need to instantiate
+        # each value from_params and return the resulting dict.
+        elif origin == Dict and len(args) == 2 and hasattr(args[-1], 'from_params'):
+            value_cls = annotation.__args__[-1]
+
+            value_dict = {}
+
+            for key, value_params in params.pop(name, Params({})).items():
+                value_dict[key] = value_cls.from_params(params=value_params, **extras)
+
+            kwargs[name] = value_dict
+
+        else:
+            # Pass it on as is and hope for the best.   ¯\_(ツ)_/¯
+            if optional:
+                kwargs[name] = params.pop(name, default)
+            else:
+                kwargs[name] = params.pop(name)
+
+    params.assert_empty(cls.__name__)
+    return {k: v for k, v in kwargs.items() if takes_arg(cls, k)}
```

Comment:
Should we really have to check for `takes_arg()` here?  When should we not crash if we get an extra one?  I _think_ the extra `Vocab` stuff has already been handled by this point, but I might be missing something.

Reply:
I think you're right